src/components/pwc-choices-2/pwc-choices-2.tsx

```diff
@@ -1,32 +1,192 @@
+
+  /**
+   * This is the mode of filtering we use to make given option objects distinct.
+   */
+  @Prop() distinctMode: PwcChoices2.DistinctMode = "none";
+  @Watch("distinctMode")
+  distinctModeWatchHandler(newValue) {
+    if (!PwcChoices2.AllDistinctModeLiterals.includes(newValue)) {
+      throwTypeLiteralNotSupported(
+        "distinctMode",
+        newValue,
+        PwcChoices2.AllDistinctModeLiterals
+      );
+    }
+  }
+
+  @Event() selectedOptionsChanged: EventEmitter<PwcChoices2.IOption[]>;
+
+  @State() selectedOptions: PwcChoices2.IOption[] = [];
+  @Watch("selectedOptions")
+  selectedOptionsWatchHandler(newValue: PwcChoices2.IOption[]) {
+    this.selectedOptionsChanged.emit(newValue);
+  }
+
+  async getSelectedOptions(mode: "option"): Promise<PwcChoices2.IOption[]>;
+  async getSelectedOptions(mode: "value" | "label"): Promise<string[]>;
+  async getSelectedOptions(
+    mode: PwcChoices2.RetreiveMode
+  ): Promise<string[] | PwcChoices2.IOption[]>;
 
-  private getText(): string {
-    return format(this.first, this.middle, this.last);
+  @Method()
+  async getSelectedOptions(
+    mode: PwcChoices2.RetreiveMode = "option"
+  ): Promise<string[] | PwcChoices2.IOption[]> {
+    switch (mode) {
+      case "option":
+        return this.selectedOptions;
+
+      case "value":
+        return this.selectedOptions.map(o => o.value);
+
+      case "option":
+        return this.selectedOptions.map(o => o.label);
+
+      default:
+        throwTypeLiteralNotSupported(
+          "mode",
+          mode,
+          PwcChoices2.AllRetreiveModeLiterals
+        );
+    }
+  }
+
+  @Listen("optionDiscarded")
+  optionDiscardedHandler(
+    event: CustomEvent<PwcChoices2.IOptionDiscardedEventPayload>
+  ) {
+    const payload = event.detail;
+    const newSelectedItems = [...this.selectedOptions];
+    newSelectedItems.splice(payload.index, 1);
+    this.selectedOptions = newSelectedItems;
+  }
+
+  @Listen("inputBarClicked")
+  inputBarClickedHandler() // event: CustomEvent<PwcChoices2.IInputBarClickedEventPayload>
+  {
+    this.dropdownIsOpen = !this.dropdownIsOpen;
+  }
+
+  @Listen("dropdownOptionClicked")
+  dropdownOptionClickedHandler(
+    event: CustomEvent<PwcChoices2.IDropdownOptionClickedEventPayload>
+  ) {
+    switch (this.type) {
+      case "multi":
+        this.selectedOptions = [...this.selectedOptions, event.detail.option];
+        break;
+      case "single":
+        this.selectedOptions = [event.detail.option];
+        break;
+    }
+  }
+
+  componentWillLoad() {
+    this.optionsWatchHandler(this.options);
+    this.typeWatchHandler(this.type);
+    this.distinctModeWatchHandler(this.distinctMode);
   }
 
   render() {
-    return <div>Hello, World! I'm {this.getText()}</div>;
+    return (
+      <div class="container">
+        {this.constructInputBar()}
+        {this.dropdownIsOpen && this.constructDropdown()}
+      </div>
+    );
+  }
+
+  constructInputBar() {
```

Comment:
`constructInputBar` and `constructDropdown` should be before the render

Reply:
fixed: c08b9a9d6c5ab24613cd90835a75fb7508d8fcd3